### PR TITLE
Simple Moving Average

### PR DESCRIPTION
**Pull request by jlinkels**
Hello Walter,

I have implemented a simple moving average for the Spectrum Analyzer. The SMA is a simple class to which values are added and a moving average is returned.

The SMA is inserted in MainFrame::DrawSpectrum. I think it belongs there and not in the Oscilloscope, I want to average the spectrum, not the display.

I have added a choice on top of the spectrum window to chose between no averaging (1) or averaging over 5, 10 or 20 spectra. (I avoid the use sample here because it is highly confusing as sampling is something else in this context). Also in the status line an indicator is added how far the averaging line is filled.

The main reason for this averaging filter is that I want to measure my subwoofer using white noise. Drawing only one spectrum does not guarantee that all frequencies have been generated during one sample period. (Meaning the audio samples here) Averaging over 20 samples gives a real nice clean picture. That is, connecting the sound card line out to the line in. I expect and got a straight line. Yet I have to measure some filters and then the subwoofer itself.

Kind regards
Johannes

### CODE REVIEW

**Comment by jlinkels** [comment]
Hello Walter,

In this commit (755bdf89ce04053464e4490bfa4338d3cc99f343) I changed some settings in the Spectrum Analyzer.

-  Moved the controls for the Spectrum Window to the side of the window as on top there was no more space to add any more controls
- Added settings for spectrum reference level (offset) and dB/div.
- Added a spectrum span of 20-2000Hz for subwoofer measurement.

For some measurements 10 dB/div is too coarse and you want 5 dB/div. And for audio 3 dB/div is convenient sometimes. When using such a scale it is almost mandatory to use a vertical offset as well to avoid the plot moves off scale. I have used "Ref level" for this offset, which is the top of the graph. This is most common with spectrum analyzers.

It would have been better to put this changes in a different branch. However, I needed to build this on top of the SMA filter to implement the new controls. I hope it still works for you.

Please let me know what you think.

Kind regards
Screenshot_20230609_010115

Johannes

**Comment by swwa** [comment]
@jlinkels Thanks, I will check it out today.

**Comment by swwa** [comment]
@jlinkels Hi Johannes,
out of curiosity what platforms are you running AudMeS on?

**Comment by jlinkels** [comment]
Hi Walter,

Thanks for looking in to my contributions.
I use AudMes only on Linux. That is, I only use Linux at all. Debian.
Only for work I have a Windows laptop.
And my IDE is vim + gdb :)

Kind regards,
Johannes

**Comment by swwa** [comment]
@jlinkels Hi Johannes,

Your contribution looks good. I have tested it on Ubuntu and Windows10. I saw some quirks when changing the buffer size but we can address these later.

I use clang-format to deal with the coding style like tabs, spaces and indentation.  Please update your changes. You can also use editorconfig plugin for vim.

Best Regards
Walter

File: sma_2d.cpp
Comment (swwa) [inline sma_2d.cpp:1]: @jlinkels Please put your name in this line.

File: sma_2d.h
Comment (swwa) [inline sma_2d.h:1]: @jlinkels Please remove the next comment lines. We do not use RCS and the copyright is on line 12

Comment (swwa) [inline sma_2d.h:1]: Unnecessary comment. Please remove

Comment (swwa) [inline sma_2d.h:1]: Ditto down to line 68.

**Review by swwa — CHANGES_REQUESTED** [review]

**Comment by jlinkels** [comment]
Hello Walter,
I hope I did make the changes according to your requests. I am pretty much a noob when it comes to contributing to someone else's github projects.
I processed all the files I worked on with clang-format and no warnings are issued anymore. Same as on the unmodified files from the master branch, so I reckoned that is correct.
Also I corrected an error in setting the average length to a fixed value of 20 when a sound interface is changed.
Please let me know if I did not do and correct I'll try again.

**Comment by jlinkels** [comment]
> Your contribution looks good. I have tested it on Ubuntu and Windows10. I saw some quirks when changing the buffer size but we can address these later.
> 
@swwa I saw the quirk after changing the buffer size  as well, and I thought I commented to you on this. But I can't find it anymore so I assume I did not.

Whenever you change the buffer size, in the original code (without averaging) the spectrum display of one buffer length goes to minimum right after changing the buffer size.

I looked into that, and to me it seems that it happens because the buffer size is expanded, and the FFT works on this expanded buffer. However, the audio sample collection is a different thread and does not know about the change in buffer size yet. So DrawSpectrum() thinks the buffer is of size m_SpeBufferLength. But it is only filled up to the old buffer size.

Then the FFT arrays realout[] en imagout[] contain zeros for the values of the frequency bins. Which is either NaN or the log10 functions return an invalid value. Actually I expected a run time error to be thrown on log10(0).

The next time audio samples are collected the buffer sizes are synchronized between the collection and the FFT and all is fine. It also only happens the first time after program start.

The difference between the original version of DrawSpectrum() and the version with the averaging inserted is that CtrlOScope seems to handle the output from log10 correctly, the SMA function does not. Therefore I added a check on the minimum value of dval AudMes.cpp:913.

In either case, for the time of one sample buffer collection the spectrum plot shows -150dB.

I did not see a quick fix for this due to the threaded and asynchronous nature of collecting the samples.

Best regards
Johannes

**Review by swwa — APPROVED** [review]
@jlinkels Thanks for your contribution!

**Pull request merged**